sql/src/main/java/io/crate/planner/operators/FetchOrEval.java

```diff
@@ -97,38 +148,62 @@ public Plan build(Planner.Context plannerContext,
+        }
 
-        if (!Symbols.containsColumn(sourceOutputs, DocSysColumns.FETCHID)) {
-            return planWithEvalProjection(plannerContext, plan, sourceOutputs);
+        if (Symbols.containsColumn(sourceOutputs, DocSysColumns.FETCHID)) {
+            return planWithFetch(plannerContext, plan, sourceOutputs);
         }
-        return planWithFetch(plannerContext, plan, sourceOutputs);
+        return planWithEvalProjection(plannerContext, plan, sourceOutputs);
     }
 
     private Plan planWithFetch(Planner.Context plannerContext, Plan plan, List<Symbol> sourceOutputs) {
         plan = Merge.ensureOnHandler(plan, plannerContext);
-        ReaderAllocations readerAllocations = plannerContext.buildReaderAllocations();
-        Map<TableIdent, FetchSource> fetchSourceByTableId = new HashMap<>(source.baseTables().size());
-        ArrayList<Reference> allFetchRefs = new ArrayList<>();
-        for (TableInfo table : source.baseTables()) {
-            if (!(table instanceof DocTableInfo)) {
-                continue;
+        Map<TableIdent, FetchSource> fetchSourceByTableId = new HashMap<>();
+        LinkedHashSet<Reference> allFetchRefs = new LinkedHashSet<>();
+
+        Map<DocTableRelation, InputColumn> fetchInputColumnsByTable = buildFetchInputColumnsMap(sourceOutputs);
+        BiConsumer<DocTableRelation, Reference> allocateFetchRef = (rel, ref) -> {
+            TableIdent tableIdent = rel.tableInfo().ident();
+            FetchSource fetchSource = fetchSourceByTableId.get(tableIdent);
+            if (fetchSource == null) {
+                fetchSource = new FetchSource(rel.tableInfo().partitionedByColumns());
+                fetchSourceByTableId.put(tableIdent, fetchSource);
+            }
+            if (ref.granularity() == RowGranularity.DOC) {
+                allFetchRefs.add(ref);
+                fetchSource.addRefToFetch(ref);
             }
-            DocTableInfo docTableInfo = (DocTableInfo) table;
-            List<Reference> fetchRefs = fetchRefsFor(outputs, docTableInfo.ident());
-            allFetchRefs.addAll(fetchRefs);
-            fetchSourceByTableId.put(
-                docTableInfo.ident(),
-                new FetchSource(
-                    docTableInfo.partitionedByColumns(),
-                    Collections.singletonList(new InputColumn(idxOfFetchId(sourceOutputs, docTableInfo.ident()))),
-                    fetchRefs
-                )
-            );
+            fetchSource.addFetchIdColumn(fetchInputColumnsByTable.get(rel));
+        };
+        List<Symbol> fetchOutputs = new ArrayList<>(outputs.size());
+        for (Symbol output : outputs) {
+            fetchOutputs.add(toInputColOrFetchRef(
+                output, sourceOutputs, fetchInputColumnsByTable, allocateFetchRef, source.expressionMapping()));
         }
+        if (source.baseTables().size() == 1) {
+            // If there are no relation boundaries involved the outputs will have contained no fields and the actions
```

Comment:
not sure I understand this.

Reply:
Could you elaborate a bit on what's unclear?

Reply:
Can you please change to `... have contained no Fields but only References ...`